### PR TITLE
26 Variable importance: support 'decision_fn' 'transform' methods

### PR DESCRIPTION
**Pull request by jpaillard**
In addition to the minor modification allowing to use more methods I added a detailed example of variable inference on a classification problem. 
It highlights: 
 - The benefits of having model-agnostic methods: here linear SVC fails when using a polynomial kernel works
 - The benefits of using conditional importance: The features are correlated making PI identify spurious important features. 
 - That the current implementation supports `decision_function` to score importance

### CODE REVIEW

**Comment by jpaillard** [comment]
The doc looks good to me. I think this is ready 
ping @dengemann

**Comment by dengemann** [comment]
Looks very cool, exactly what I had in mind!

File: hidimstat/cpi.py [hunk 1]
```diff
@@ -170,10 +173,7 @@ def _joblib_predict_one_group(imputation_model, X, j):
                 if isinstance(X, pd.DataFrame):
                     X_perm = pd.DataFrame(X_perm, columns=X.columns)
 
-                if self.score_proba:
-                    y_pred_perm = self.estimator.predict_proba(X_perm)
-                else:
-                    y_pred_perm = self.estimator.predict(X_perm)
+                y_pred_perm = getattr(self.estimator, self.method)(X_perm)
```
Comment (dengemann) [inline hidimstat/cpi.py:176]: probably this should be handled explicitly

```Python

y_pred_perm = getattr(self.estimator, self.method, None)(X_perm)
if y_pred_perm is None:
    raise ValueError(r"The predict method ({self.method}) you passed is not supported by the estimator")
```

File: hidimstat/cpi.py [hunk 2]
```diff
@@ -217,10 +217,7 @@ def score(self, X, y):
 
         out_dict = dict()
 
-        if self.score_proba:
-            y_pred = self.estimator.predict_proba(X)
-        else:
-            y_pred = self.estimator.predict(X)
+        y_pred = getattr(self.estimator, self.method)(X)
```
Comment (dengemann) [inline hidimstat/cpi.py:220]: also here. perhaps factorize it into a small getter function

File: hidimstat/cpi.py (hunk 1 above)
Comment (dengemann) [reply hidimstat/cpi.py:176]: ok. Then fine!

**Comment by jpaillard** [comment]
ping @bthirion

**Review by bthirion — COMMENTED** [review]
Sorry, I had done the review a few days ago but forgot to submit it ...

**Comment by jpaillard** [comment]
I added the test for Errors. 
This is ready on my side

File: examples/plot_variable_importance_classif.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+model_non_linear = RandomizedSearchCV(
+    SVC(
+        random_state=seed,
+        kernel="poly",
+        degree=2,
+        coef0=1,
+    ),
+    param_distributions={
+        "C": np.logspace(-3, 3, 10),
+        # "gamma": np.logspace(-3, 3, 10),
```
Comment (bthirion) [inline examples/plot_variable_importance_classif.py:149]: remove commented lines

File: examples/plot_variable_importance_classif.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+        "Age",
+        "X8",
+        "X9",
+        "X10",
+    ]
+)
+tck_ids = np.arange(0, n_features)
+axes[1].set_xticks(tck_ids, labels[tck_ids], rotation=45)
+axes[1].set_yticks(tck_ids, labels[tck_ids])
+
```
Comment (bthirion) [inline examples/plot_variable_importance_classif.py:114]: What do the red/blue dots correspond to ? Can you add a legend ?

File: examples/plot_variable_importance_classif.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        diff = importance_arr[:, i, j]
+        pval_arr[j, i] = ttest_1samp(diff, 0)[1]
+
+#############################################################################
+# Visualize the variable importance
+# ------------------------------
+# Here we plot the variable importance and highlight the features that are considered
+# important, with a p-value lower than 0.05, using a diamond marker. We also highlight
+# the true important features, used to generate the target variable, with a star marker.
+# While the linear model captures the importance of the age, it fails to capture the the
```
Comment (bthirion) [inline examples/plot_variable_importance_classif.py:226]:
```suggestion
# While the linear model captures the importance of the age, it fails to capture the
```

**Review by bthirion — COMMENTED** [review]
Minor improvements needed, but this looks very good overall !

File: examples/plot_variable_importance_classif.py (hunk 4 above)
Comment (jpaillard) [reply examples/plot_variable_importance_classif.py:114]: I updated the legend, it corresponds to variables for which the p-value is < threshold (here I used 0.05)

**Comment by jpaillard** [comment]
I addressed your comments @bthirion

File: hidimstat/cpi.py [hunk 6]
```diff
@@ -42,7 +44,7 @@ def __init__(
         imputation_model,
         n_permutations: int = 50,
         loss: callable = root_mean_squared_error,
-        score_proba: bool = False,
+        method: str = "predict",
```
Comment (bthirion) [inline hidimstat/cpi.py:47]: You need to update the docstring above, as well as for LOCO and PI.

**Review by bthirion — COMMENTED** [review]
Could you just update the docstrings ? Then we can merge.
Thx,

File: hidimstat/cpi.py (hunk 6 above)
Comment (jpaillard) [reply hidimstat/cpi.py:50]: done

**Review by bthirion — APPROVED** [review]

**Comment by bthirion** [comment]
Merging now, thx.

**Pull request merged**